### PR TITLE
Added extra validation for dates

### LINKED ISSUE #403: Better validation error when date is invalid
Currently we just use '... is required' for all validation messages. If the date is invalid because it's impossible we should adjust our error message accordingly

Should handle these cases:
 - A date which is out of range or has invalid characters (what should a valid range be?)
 - An end date which is before the begin date
 - A birthdate which is after today (do we want to restrict it further to at least 13 years old?)

### PR DESCRIPTION
**Pull request by noisecapella**
#### What are the relevant tickets?
Fixes #403 
Rebased on #504 

#### What's this PR do?
Handles the three validation cases mentioned in the issue

#### How should this be manually tested?
 - Verify that an error occurs if you change your date of birth to a date in the future and click Save
 - In a work history entry try to add a year but not a month for the end date. Verify that the error message is: "Please enter a valid end date or leave it blank". If you remove both year and month, everything should work fine since that means you are currently working there.
 - In a work history entry verify that you get an error if your end date is before your start date. Verify that the message is: "End date cannot be before start date"
 - Also in work history, remove the start date, click 'Save' and verify that the error message is "Please enter a valid start date"
 - In an education entry verify that you see this error message when the graduation date is missing or invalid: "Please enter a valid graduation date"

### CODE REVIEW

**Comment by aliceriot** [comment]
code looks good, testing functionality

**Comment by aliceriot** [comment]
- [x] Verify that an error occurs if you change your date of birth to a date in the future and click Save
- [ ] In a work history entry try to add a year but not a month for the end date. Verify that the error message is: "Please enter a valid end date or leave it blank". If you remove both year and month, everything should work fine since that means you are currently working there.
- [ ] In a work history entry verify that you get an error if your end date is before your start date. Verify that the message is: "End date cannot be before start date"
- [ ] Also in work history, remove the start date, click 'Save' and verify that the error message is "Please enter a valid start date"
- [ ] In an education entry verify that you see this error message when the graduation date is missing or invalid: "Please enter a valid graduation date"

**Comment by aliceriot** [comment]
for the second case I found an edge case.

1. create a work history entry that has a proper end date
2. edit that work history
3. delete the 'month' field value
4. click 'save'
5. entry is saved without an end date (despite there being a year value in the year field when 'save' is clicked)

**Comment by aliceriot** [comment]
one edge case, other than that things seem to be working correctly.

**Comment by noisecapella** [comment]
Good catch, it looks like `_.isEmpty(number)` returns true... :see_no_evil:

**Comment by noisecapella** [comment]
Ok, that case should be handled now. Also, I discovered that moment JS parses two digit years 19xx despite being passed a format string with YYYY, so I changed the year validation to require all dates to be within 1800 >= year > 2100

**Comment by aliceriot** [comment]
Yep, confirmed that case is fixed! looking over the code now.

File: static/js/util/validation.js [hunk 1]
```diff
@@ -11,12 +11,17 @@ let handleNestedValidation = (profile, keys, nestedKey) => {
   return _.flatten(profile[nestedKey].map((v, i) => nestedFields(i)));
 };
 
+/**
+ * Is the value === undefined, null, or ""?
+ */
```
Comment (aliceriot) [inline static/js/util/validation.js]: I don't think this comment is super necessary, I think this function documents itself. Personal style preference though.

**Comment by aliceriot** [comment]
One little comment, other than that I think the code looks :100:

**Comment by noisecapella** [comment]
Comment addressed

**Comment by aliceriot** [comment]
it's a :+1: from me!

**Pull request merged**